### DISCUSSION TITLE
ViewHosting Question

### DISCUSSION THREAD

**Discussion by gili-labs** [nalexn/ViewInspector#91] [Q&A]
Why doesn't ViewHosting.host(view:) cause a window to appear on the display?

**Comment by nalexn** [comment]
Hey @gili-labs , I didn't explore why, is there a use case for why it should appear?

**Reply by gili-labs** [reply]
I'm trying to figure out a way to simulate dragging.

As I continued down this path, I became curious why a window doesn't show up after ViewHosting.host(view:) has executed. I have set a breakpoint inside didAppear(_:) and looked around for the window, but don't see it.

**Reply by gili-labs** [reply]
I'm taking a different path, exploring what you did for onTapGesture and onLongPressGesture. All I need is a way to call one of the registered callbacks, delivering a DragGesture.Value, which I can set to anything I want. This would suffice to simulate a drag gesture. Might take a bit of work though, as I'm trying to unravel what you did and it isn't straightforward. Are there any tools built into the library that might help me identify the necessary modifier?

**Reply by nalexn** [reply]
Hey @gili-labs , yea, let me give you a few tips. 

Here is how I usually approach further development:

1. In the test scope I add a simple test view that uses the modifier I'm interested in. In your example, it could be

```swift
@available(iOS 13.0, macOS 10.15, tvOS 13.0, *)
struct TestView: View, Inspectable {
    
    @State var offset: CGFloat = 0
    
    var body: some View {
        Text("abc")
            .offset(x: 0, y: offset)
            .gesture(
                DragGesture()
                    .onChanged { gesture in
                        self.offset = gesture.translation.height
                    }
                    .onEnded { _ in
                        self.offset = .zero
                    }
            )
    }
}
```

2. I add a dummy test like so:

```swift
func testNewModifier() throws {
    let sut = TestView()
    _ = sut
}
```

and put a breakpoint after sut.

3. I run the test, and when debugger stops on the breakpoint, I go to lldb and execute `po Inspector.print(sut) as AnyObject`

It prints the entire structure of the view, which gives me all the info needed.

---

Going back to your example, Gestures are quite a big chunk of work. That's because gestures are not a simple modifier, but whey form its own hierarchy with special modifiers.

You can run `po Inspector.print(...) as AnyObject` for any value, so try running this for the following:

```swift
let gesture1 = DragGesture()
let gesture2 = DragGesture()
            .onChanged { _ in }
            .onEnded { _ in }
```

You'll see that there is a base structure that gets wrapped in modifiers just like Views are. The good news is that all the callbacks are accessible (according to the print's output), so simulating gestures is doable.

So, in order to do achieve what you want, we'd need to do the following:

1. Introduce a separate infrastructure for unwrapping Gestures from their modifiers
2. Declare base types for all the standard gesture types so you could unwrap a specific one
3. Implement a function that would find the right modifier applied to the gesture and call the callback (`onChanged` for example). Providing the correct parameters for the closure sometimes is trivial, sometimes is quite challenging
4. Implement inspection call for `gesture` view modifier that would give access to inspecting the Gesture with its modifiers

As you can see, there is a huge amount of work for such a simple task, probably that's why I deferred working on it.
I really appreciate your willingness to contribute and I would not expect anyone to jump right into such a complex case, so let me know if you do want to proceed with this one or have a practice on a simpler modifier that's not supported yet by the library. I'm happy to answer any questions!

**Reply by gili-labs** [reply]
I'd like to take a stab at it. I'm enjoying work with this library, as it's a wonderful work of art. I often scratch my head and ask myself why Apple makes it so difficult for us to test our applications and frameworks. With that said, I think UI testing is a horrible, brute force approach to testing, and with the advent of a declarative framework like SwiftUI, I see even less point.

On another note, have you had a chance to review the pull request I submitted last week?

**Reply by nalexn** [reply]
Yea, sorry for not merging it in yet, I'm going to release a new version this week that'll include your PR.

**Reply by gili-labs** [reply]
No sweat, I was just curious.

**Reply by gili-labs** [reply]
I just finished the implementation and tests covering DragGesture.onChanged(_:) and DragGesture.onEnded(_:). I'm going to take a crack at DragGesture.updating(_:body:). I also want to test if they work with DragGesture.simultaneously(with:), DragGesture.sequenced(before:), and DragGesture.exclusively(before:). If I'm feeling ambitious, I may take DragGesture.modifiers(_:) and DragGesture.map(_:); however, I have never used these and I'm going to have to determine what use cases they serve.

**Reply by gili-labs** [reply]
Testing of `DragGesture` went well. However, when the `DragGesture` is part of a composed gesture, it fell apart. As I opened this can of worms, I saw two problems complicating matters:
- Composed gestures enable an application to create a hierarchy of gestures.
- The `Gesture` protocol enables an application to define a custom gesture.

Hence, while it easy enough to extract a gesture added by the `gesture(_:)` modifier, this may result a single gesture or a tree of gestures, SwiftUI may define the gesture, or the application may have defined it.

On the upside, the fact that all gestures conform to the `Gesture` protocol simplifies the problem slightly. However, I can't help thinking that a proper unit testing framework supporting gestures allows for the inspection of a gesture, including its attributes and the ability to invoke its actions. To do this, the framework would essentially have to make Gesture inspectable, not unlike it makes View inspectable.

I'm pouring over your code trying to determine whether it is possible to make gestures inspectable using existing code, or would it require duplication of some of the code with the intent of supporting gestures, rather than views and view modifiers.

**Reply by nalexn** [reply]
Yeah, I see what you're talking about. I think the way to approach this can be to have all your working code reside in `extension Gesture where Self: Inspectable` extensions, then for all the standard gestures add such conformance: `extension DragGesture: Inspectable { }`, and finally expect the user to add conformance for their custom gestures.

**Reply by gili-labs** [reply]
Brilliant! Let me give that a try and see where it leads me.

**Reply by gili-labs** [reply]
@nalexn, I wanted to provide you with a quick update on generic gesture support. I have made some progress. However, the nature of gestures have challenged me to reexamine my approach more than once. I am taking some vacation next week, during which time I hope to finish this up.

**Reply by nalexn** [reply]
@gili-labs sounds good 👍

**Reply by gili-labs** [reply]
Finished support for all gestures defined by SwiftUI, including composed gestures. I need to add support for custom gestures, clean up the code a bit, and write documentation. I might have something by the end of this upcoming weekend.

**Reply by gili-labs** [reply]
I started at looking at custom gestures, and I don't mean composed gestures (which many people think are custom gestures). Rather, I tried to implement a structure that conforms to the `Gesture` protocol. I cannot figure this out. This protocol starts out by defining the following:

```Swift
public protocol Gesture {

    /// The type representing the gesture's value.
    associatedtype Value

    /// The type of gesture representing the body of `Self`.
    associatedtype Body : Gesture

    /// The content and behavior of the gesture.
    var body: Self.Body { get }
}
```

This seems circular to me. How can I define a structure with a property having type `Body`, when the `Body` itself must conform to the `Gesture` protocol? Do you have any ideas? Maybe Apple has some hidden structures that define Gesture types, the guts that actually implement the gesture recognizer. I've Googled this and it doesn't seem anyone has figured out how to define a custom gesture to encapsulate custom gesture recognizers.

If I cannot define a custom gesture, then I don't see the point in implementing support for custom gestures.

**Reply by nalexn** [reply]
@gili-labs I believe `associatedtype Body : Gesture` is not constrained to be `Self`, which would indeed mean a circular reference. It's another gesture type, a composed one, for example.
The `View` protocol declaration looks similar.

**Reply by gili-labs** [reply]
@nalexn Given your response, I was able to finesse a custom gesture. It doesn't make sense to me, but at least it's enough to create a vehicle for unit testing.

Here's what I don't understand. The `Gesture` protocol requires a property called body of type `Gesture`. Of course, the property cannot have the type `Self`, since this would introduce a recursive storage requirement, which the compiler doesn't allow. However, this assumes that at some point there is some basic types that conform to `Gesture` that new gestures can draw on. I was a bum and simply did this:

```Swift
struct CustomGesture: Gesture {
    typealias Value = ()
    
    var body: TapGesture
}
```

However, I'm scratching my head wondering how the SwiftUI team defined the set of gestures that come with SwiftUI. I see there is a similar head scratching exercise with types that conform to `View`. Is there some kind of magic going on behind the curtains, or am I simply overlooking something?

**Reply by nalexn** [reply]
Apparently, Apple's using private APIs for writing the root View and Gesture types, so we cannot introduce our own, but only utilize the existing set. If you think about this - there is no access to the lower level framework from SwiftUI: in UIKit you could write a completely custom view using CoreAnimation/CoreGraphics, or OpenGLES, but not with SwiftUI. I think this is done intentionally, so that Apple could replace the UIKit that is currently powering the SwiftUI with direct rendering on Metal, for example.

**Reply by gili-labs** [reply]
As I thought about this more, I became curious about `EmptyView` and found that its `Body` is `Never`. Hence, I decided to try to define a gesture with a `Body` equal to `Never`. The compiler complained that the CustomGesture did not conform to Gesture and would I like it to add the necessary stubs, which I opted for. Look what it added:

```Swift
struct CustomGesture: Gesture {
    static func _makeGesture(gesture: _GraphValue<CustomGesture>, inputs: _GestureInputs) -> _GestureOutputs<()> {
        <#code#>
    }
    
    typealias Value = ()
    
    var body: Never
}
```

So, there is some undocumented goodies. Unfortunately, this uses a number of undocumented types. It leads me to think that Apple is supporting "custom gestures" at this time. Apple must feel that we can do anything through gesture composition, which I don't necessarily agree with.

**Reply by nalexn** [reply]
Haha, good find! Yeah, there are quite a lot of these APIs starting with "_" that are visible to the compiler but get trimmed out when you look at the framework header file in Xcode. Whenever you see this underscored API you can be sure Apple doesn't want you to use it in the client code directly. So yeah, you found the backdoor for writing the custom gesture, but that's not the way they intended us to do it.

**Reply by gili-labs** [reply]
@nalexn do you know a handful of unit tests don't master in the current commit to master?

**Reply by nalexn** [reply]
@gili-labs I'm unaware of that. CI is broken right now (Travis has become paid), so I need to migrate to Github actions, but I do run `pod lint` locally that runs tests suit against all platforms. What are the errors?

**Reply by gili-labs** [reply]
@nalexn I run the units tests quite often and they've been alright, but I was a bit behind and had to merge latest. After I pulled, I ran the tests again and noticed errors. I automatically assumed it was something I did. So, I cloned directly from your repo, built master (none of the new gesture code), and ran the unit tests. Here is a list of errors:

In TextTests:

```
Test Case '-[ViewInspectorTests.TextTests testLocalizationGlobalDefault]' started.
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/SwiftUI/TextTests.swift:55: error: -[ViewInspectorTests.TextTests testLocalizationGlobalDefault] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/SwiftUI/TextTests.swift:60: error: -[ViewInspectorTests.TextTests testLocalizationGlobalDefault] : XCTAssertEqual failed: ("Test") is not equal to ("Test_en")
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/SwiftUI/TextTests.swift:65: error: -[ViewInspectorTests.TextTests testLocalizationGlobalDefault] : XCTAssertEqual failed: ("Test") is not equal to ("Тест_ru")
Test Case '-[ViewInspectorTests.TextTests testLocalizationGlobalDefault]' failed (0.315 seconds).
Test Case '-[ViewInspectorTests.TextTests testObjectInitialization]' started.
```

```
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/SwiftUI/TextTests.swift:41: error: -[ViewInspectorTests.TextTests testResourceLocalizationStringNoParams] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/SwiftUI/TextTests.swift:47: error: -[ViewInspectorTests.TextTests testResourceLocalizationStringNoParams] : XCTAssertEqual failed: ("Test") is not equal to ("Test_en")
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/SwiftUI/TextTests.swift:49: error: -[ViewInspectorTests.TextTests testResourceLocalizationStringNoParams] : XCTAssertEqual failed: ("Test") is not equal to ("Test_en_au")
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/SwiftUI/TextTests.swift:51: error: -[ViewInspectorTests.TextTests testResourceLocalizationStringNoParams] : XCTAssertEqual failed: ("Test") is not equal to ("Тест_ru")
Test Case '-[ViewInspectorTests.TextTests testResourceLocalizationStringNoParams]' failed (0.068 seconds).
Test Case '-[ViewInspectorTests.TextTests testSearch]' started.
Test Case '-[ViewInspectorTests.TextTests testSearch]' passed (0.002 seconds).
```

In ViewSearchTests:

```
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:91: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : XCTAssertEqual failed: ("["Btn", "Test", "123", "xyz", "Modifier"]") is not equal to ("["Btn", "Test_en", "123", "xyz", "Modifier"]")
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindAll] : failed - Failed to load test Bundle
Test Case '-[ViewInspectorTests.ViewSearchTests testFindAll]' failed (0.094 seconds).
Test Case '-[ViewInspectorTests.ViewSearchTests testFindButton]' started.
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindButton] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindButton] : failed - Failed to load test Bundle
```

```
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindButton] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindButton] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindButton] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindButton] : failed - Failed to load test Bundle
Test Case '-[ViewInspectorTests.ViewSearchTests testFindButton]' failed (0.033 seconds).
Test Case '-[ViewInspectorTests.ViewSearchTests testFindCustomView]' started.
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindCustomView] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindCustomView] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindCustomView] : failed - Failed to load test Bundle
```

```
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindCustomView] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindCustomView] : failed - Failed to load test Bundle
Test Case '-[ViewInspectorTests.ViewSearchTests testFindCustomView]' failed (0.032 seconds).
Test Case '-[ViewInspectorTests.ViewSearchTests testFindForConditionalView]' started.
Test Case '-[ViewInspectorTests.ViewSearchTests testFindForConditionalView]' passed (0.004 seconds).
Test Case '-[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithGlobalDefault]' started.
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithGlobalDefault] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithGlobalDefault] : failed - Failed to load test Bundle
```

```
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/BaseTypesTests.swift:40: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Expression did not throw any error
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
```

```
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:134: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : XCTAssertNoThrow failed: threw error "Search did not find a match"
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:135: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : XCTAssertNoThrow failed: threw error "Search did not find a match"
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:137: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : XCTAssertNoThrow failed: threw error "Search did not find a match"
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:139: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : XCTAssertNoThrow failed: threw error "Search did not find a match"
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter] : failed - Failed to load test Bundle
Test Case '-[ViewInspectorTests.ViewSearchTests testFindLocalizedTextWithLocaleParameter]' failed (0.098 seconds).
Test Case '-[ViewInspectorTests.ViewSearchTests testFindMatchingBlockerView]' started.
Test Case '-[ViewInspectorTests.ViewSearchTests testFindMatchingBlockerView]' passed (0.002 seconds).
Test Case '-[ViewInspectorTests.ViewSearchTests testFindText]' started.
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindText] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindText] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:101: error: -[ViewInspectorTests.ViewSearchTests testFindText] : XCTAssertEqual failed: threw error "Search did not find a match"
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindText] : failed - Failed to load test Bundle
```

```
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindText] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindText] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindText] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindText] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindText] : failed - Failed to load test Bundle
Test Case '-[ViewInspectorTests.ViewSearchTests testFindText]' failed (0.071 seconds).
Test Case '-[ViewInspectorTests.ViewSearchTests testFindViewWithId]' started.
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindViewWithId] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindViewWithId] : failed - Failed to load test Bundle
Test Case '-[ViewInspectorTests.ViewSearchTests testFindViewWithId]' failed (0.017 seconds).
Test Case '-[ViewInspectorTests.ViewSearchTests testFindViewWithTag]' started.
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindViewWithTag] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindViewWithTag] : failed - Failed to load test Bundle
/Users/patrick/Documents/Apps/SwiftUI/tmpvi/ViewInspector/Tests/ViewInspectorTests/ViewSearchTests.swift:14: error: -[ViewInspectorTests.ViewSearchTests testFindViewWithTag] : failed - Failed to load test Bundle
Test Case '-[ViewInspectorTests.ViewSearchTests testFindViewWithTag]' failed (0.018 seconds).
Test Case '-[ViewInspectorTests.ViewSearchTests testStubViewBody]' started.
```

```
Test Case '-[ViewInspectorTests.ViewSearchTests testStubViewBody]' passed (0.000 seconds).
Test Suite 'ViewSearchTests' failed at 2021-04-08 10:56:03.609.
	 Executed 12 tests, with 62 failures (1 unexpected) in 0.390 (0.393) seconds
Test Suite 'ViewSizingTests' started at 2021-04-08 10:56:03.609
Test Case '-[ViewInspectorTests.ViewSizingTests testFixedSize]' started.

Test Case '-[ViewInspectorTests.ViewSearchTests testStubViewBody]' passed (0.000 seconds).
Test Suite 'ViewSearchTests' failed at 2021-04-08 10:56:03.609.
	 Executed 12 tests, with 62 failures (1 unexpected) in 0.390 (0.393) seconds
Test Suite 'ViewSizingTests' started at 2021-04-08 10:56:03.609
Test Case '-[ViewInspectorTests.ViewSizingTests testFixedSize]' started.
```

**Reply by gili-labs** [reply]
For snicks and giggles, I poked around and for TextTests I was able to make it work by doing the following:

```Swift
@available(iOS 13.0, macOS 10.15, tvOS 13.0, *)
extension Bundle {
    
    static func testResources(file: StaticString = #file, line: UInt = #line) throws -> Bundle {
//        let bundle = Bundle(for: TextTests.self)
        let bundle = Bundle.module
        guard bundle.path(forResource: "Test", ofType: "strings") != nil else {
            XCTFail("Failed to load test Bundle", file: file, line: line)
            return .main
        }
        return bundle
    }
```

I imagine similar problems reside in the other tests. I'll keep poking around.

**Reply by nalexn** [reply]
Hm. For me it says `Type 'Bundle' has no member 'module'`. Are you running a beta Xcode? I'm on 12.4

**Reply by gili-labs** [reply]
Very odd...I'm running Version 12.4 (12D4e). I only run beta from WWDC to first GA, because I'm some kind of a masocist.

**Reply by nalexn** [reply]
Ok, I figured this out. I'm opening `ViewInspector.xcodeproj` and running tests from there, you're opening and running  `Package.swift`. `Bundle.module` is not available when running from under `ViewInspector.xcodeproj`, and I do see the tests failing in your setup. I'll have a look if I can resolve this

**Reply by nalexn** [reply]
Pushed a fix to master

**Reply by gili-labs** [reply]
Cool, I'll pull and run tests again in my branch. All I have after this is documentation and I submit a pull request. I'm still forecasting by the end of this weekend.

**Comment by jsoneaday** [comment]
What does ViewHosting actually mean? Its not clear to me.

**Reply by nalexn** [reply]
Some SwiftUI state entities (such as `@State` or Environment) do not gain access to the values store until they are added to the render routine (displayed on screen). `ViewHosting` is used in [async tests](https://github.com/nalexn/ViewInspector/blob/master/guide.md#views-using-state-environment-or-environmentobject).

**Comment by sisoje** [comment]
> Why doesn't ViewHosting.host(view:) cause a window to appear on the display?

If you want it to display on the screen then you need to host the view in a real app. You can use the unofficial approach I made here: https://github.com/sisoje/testable-view-swiftui
